### PR TITLE
Fix hashing when objects are wrapped in primitives

### PR DESCRIPTION
**Pull request by yashgorana**
## Description
`List[SyftObject]`, `Dict[str, SyftObject]` or any other primitive types  that used `recursive_serde_register` was not being hashed correctly. The is fixed by detecting, caching and passing-through args/kwargs when `sy.serialize()` **while in recursion**.

Discovered as a part of https://github.com/OpenMined/Heartbeat/issues/324 & relates to https://github.com/OpenMined/PySyft/pull/7732

## Affected Dependencies
- All objects using `sy.serialize()`

## How has this been tested?
- Update unit tests with appropriate scenarios

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

**Comment by yashgorana** [comment]
@madhavajay @rasswanth-s this is now fixed. Please see the description as to why the fix had to be done in this way

**Comment by yashgorana** [comment]
@rasswanth-s @koenvanderveen @shubham3121 Madhava asked me to have this reviewed by you folks, can you please take a look at this review? I've updated this change to be a more cleaner + robust in multi-threaded scenarios (syft running with threads through gunicorn)

File: packages/syft/src/syft/serde/serialize.py [hunk 1]
```diff
@@ -1,20 +1,87 @@
...
+        if ctx.in_recursion:
+            # select start args will be passed to recursive serde invocations
+            start_args = ctx.get("start_args")
+            for_hashing = start_args.get("for_hashing")
+            debug = start_args.get("debug")
+        else:
+            # save the start args in serde context
+            start_args = dict(for_hashing=for_hashing, debug=debug)
+            ctx.set("start_args", start_args)
+
```
Comment (yashgorana) [inline packages/syft/src/syft/serde/serialize.py:74]: The is fixed by detecting, caching and passing-through args/kwargs when sy.serialize() while **in recursion**.  `RecursiveSerdeContext` detects the recursion and provides a thread-specific local storage where I save the args/kwargs. Rest is just pass-through code.

File: packages/syft/src/syft/serde/serialize.py [hunk 2]
```diff
@@ -1,20 +1,87 @@
...
+        if debug:
+            print(
+                f'{"-" * ctx.recurse_step}>',
+                f"Obj = {obj.__class__.__name__}@{hex(id(obj))} | Value = {obj}",
+            )
```
Comment (yashgorana) [inline packages/syft/src/syft/serde/serialize.py:79]: Sample debug output

```
-> Obj = MockNestedList@0x7f92ad796940 | Value = __main__.MockNestedList
--> Obj = list@0x7f92adae4500 | Value = [__main__.MockSyftObject, __main__.MockSyftObject, __main__.MockSyftObject]
---> Obj = MockSyftObject@0x7f92ad796ac0 | Value = __main__.MockSyftObject
----> Obj = UID@0x7f92ad776320 | Value = 99328247805c4bd18cb5dc7388b3bef9
-----> Obj = bytes@0x7f92ad7c1eb0 | Value = b'\x992\x82G\x80\\K\xd1\x8c\xb5\xdcs\x88\xb3\xbe\xf9'
---> Obj = MockSyftObject@0x7f92ad796b00 | Value = __main__.MockSyftObject
----> Obj = UID@0x7f92ad776410 | Value = 4e615d65429e4573911346e66cdcd1b6
-----> Obj = bytes@0x7f92ad7c1ff0 | Value = b'Na]eB\x9eEs\x91\x13F\xe6l\xdc\xd1\xb6'
---> Obj = MockSyftObject@0x7f92ad796b40 | Value = __main__.MockSyftObject
----> Obj = UID@0x7f92ad776500 | Value = 8142ae58cd42408691900d9fe5f998bf
-----> Obj = bytes@0x7f92ad7c2070 | Value = b'\x81B\xaeX\xcdB@\x86\x91\x90\r\x9f\xe5\xf9\x98\xbf'
--> Obj = UID@0x7f92d35836d0 | Value = 572feb9669204a8cbb41a0ab4cbf0093
---> Obj = bytes@0x7f92ad7c0830 | Value = b'W/\xeb\x96i J\x8c\xbbA\xa0\xabL\xbf\x00\x93'

CPU times: user 404 µs, sys: 0 ns, total: 404 µs
Wall time: 438 µs
```

File: packages/syft/tests/syft/hash_test.py [hunk 3]
```diff
@@ -29,6 +36,35 @@ class MockWrapper(SyftBaseObject, SyftHashableObject):
...
+    __canonical_name__ = "MockSyftObject"
+    __version__ = SYFT_OBJECT_VERSION_1
+
+
+@serializable()
+class MockNestedList(SyftObject):
+    __canonical_name__ = "TestNestedList"
+    __version__ = SYFT_OBJECT_VERSION_1
+
+    classes: List[MockSyftObject]
```
Comment (koenvanderveen) [inline packages/syft/tests/syft/hash_test.py:50]: why is this called `classes`, doesnt this just contain the underlying data?

File: packages/syft/src/syft/serde/serialize.py [hunk 4]
```diff
@@ -1,20 +1,87 @@
...
+from typing_extensions import Self
+
+
+class RecursiveSerdeContext:
+    """Context manager singleton that tracks recursive SerDe calls in a thread-safe way"""
+
+    _instance: Optional[Self] = None
+    _store = threading.local()
+
+    def __new__(cls) -> Any:
```
Comment (koenvanderveen) [inline packages/syft/src/syft/serde/serialize.py:16]: 👍

**Comment by koenvanderveen** [comment]
I am not completely understanding the problem this solves. If I remove the lines with the context, the tests are still passing

**Comment by madhavajay** [comment]
@rasswanth-s and @yashgorana can we discuss this again in slack and get it merged if its working and tested?

**Pull request closed without merging**